publisher-sdk/src/androidTest/java/com/criteo/publisher/advancednative/CriteoImageLoaderTest.java

```diff
@@ -0,0 +1,68 @@
+/*
+ *    Copyright 2020 Criteo
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.criteo.publisher.advancednative;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import android.content.Context;
+import android.view.LayoutInflater;
+import android.widget.ImageView;
+import com.criteo.publisher.DependencyProvider;
+import com.criteo.publisher.concurrent.ThreadingUtil;
+import com.criteo.publisher.concurrent.TrackingCommandsExecutor;
+import com.criteo.publisher.tests.R;
+import com.criteo.publisher.mock.MockedDependenciesRule;
+import java.net.MalformedURLException;
+import java.net.URL;
+import javax.inject.Inject;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class CriteoImageLoaderTest {
+
+  @Rule
+  public MockedDependenciesRule mockedDependenciesRule = new MockedDependenciesRule();
+
+  @Inject
+  private Context context;
+
+  @Test
+  public void givenKittenUrl_ImageLoader_LoadsIt() throws Exception {
+    ThreadingUtil.runOnMainThreadAndWait(() -> {
+      LayoutInflater inflater = LayoutInflater.from(context);
+      ImageView imageView = (ImageView) inflater.inflate(R.layout.test_image_loader, null);
+
+      DependencyProvider dependencyProvider = mockedDependenciesRule.getDependencyProvider();
+      CriteoImageLoader imageLoader = (CriteoImageLoader) dependencyProvider
+          .provideDefaultImageLoader();
+
+      URL url = null;
+      try {
+        url = new URL("https://placekitten.com/300/500");
+      } catch (MalformedURLException e) {
+        // no-op
+      }
+      assertThat(url).isNotNull();
+      imageLoader.loadImageInto(url, imageView, null);
+      TrackingCommandsExecutor trackingCommandsExecutor = (TrackingCommandsExecutor) dependencyProvider
+          .provideThreadPoolExecutor();
+
+      ThreadingUtil.waitForAllThreads(trackingCommandsExecutor);
```

Comment:
I guess that here you cannot use the `mockedDependencyRule.waitForIdleState()` because you are already on the main thread, which is managed. Using the `waitForAllThreads` requires you to cast the thread pool, and implicitly know that idle state is waited this way.

Personally, I prefer to stay mostly on the test thread and jump on UI thread only for precise operations. You have two tools for this:
- `runOnMainThreadAndWait` to execute an action (using `Runnable`)
- `callOnMainThreadAndWait` to execute a function returning a value (using `Callable<V>`)

In this test you need to be twice on the UI thread:
- to instantiate the `ImageView` via `callOnMainThreadAndWait`
- to load the image into the image view via `runOnMainThreadAndWait`

I propose you to use UI thread only for this precise operations. Then you won't need to manually handle exception, manually handle idle state.

Reply:
Done. That said, `Exception` would have to be handled if I use an `ImageLoader` instead of `CriteoImageLoader` as `ImageLoader` method signatures throw `Exceptions`.

Reply:
You might use the `Callable` version then. `Callable` throws exception. Then you have to return `null`.